psi4/src/export_mints.cc

```diff
@@ -225,32 +225,57 @@ std::shared_ptr<Molecule> from_dict(py::dict molrec) {
     std::vector <std::string> elem = molrec["elem"].cast<std::vector <std::string>>();
     std::vector <double> mass = molrec["mass"].cast<std::vector <double>>();
     std::vector <int> real = molrec["real"].cast<std::vector <int>>();
-    std::vector <std::string> elbl= molrec["elbl"].cast<std::vector <std::string>>();
-    size_t nat = geom.size() / 3;
-
-    for (size_t iat=0; iat<nat; ++iat) {
-        std::string symbol = elem[iat];
-        std::transform(symbol.begin(), symbol.end(), symbol.begin(), ::toupper);
-        mol->add_atom(elez[iat] * real[iat],
-                      geom[3*iat], geom[3*iat+1], geom[3*iat+2],
-                      symbol,
-                      mass[iat],
-                      elez[iat] * real[iat],
-                      symbol + elbl[iat],
-                      elea[iat]);
-    }
+    std::vector <std::string> elbl = molrec["elbl"].cast<std::vector <std::string>>();
+
+    size_t nat;
+    bool unsettled;
+    if (_has_key(molrec, "geom_unsettled")) {
+        std::vector <std::vector <std::string>> geom_unsettled = molrec["geom_unsettled"].cast<std::vector <std::vector <std::string>>>();
+        nat = geom_unsettled.size();
+        unsettled = true;
+
+        for (size_t iat=0; iat<nat; ++iat) {
+            std::string symbol = elem[iat];
+            std::string label = elbl[iat];
+            std::transform(symbol.begin(), symbol.end(), symbol.begin(), ::toupper);
+            std::transform(label.begin(), label.end(), label.begin(), ::toupper);
+            mol->add_unsettled_atom(elez[iat] * real[iat],
+                                    geom_unsettled[iat],
+                                    symbol,
+                                    mass[iat],
+                                    elez[iat] * real[iat],
+                                    symbol + label,
+                                    elea[iat]);
+        }
 
-    ////    mol->set_has_zmatrix(false);  // TODO
-    ////    moldict['zmat'] = self.zmat
-    ////    moldict['reinterpret_coordentries'] = self.PYreinterpret_coordentries
-    ////    moldict['lock_frame'] = self.lock_frame
-    ////    mol->set_reinterpret_coordentry(false);
+        std::vector <std::pair <std::string, double>> variables = molrec["variables"].cast<std::vector <std::pair <std::string, double>>>();
+        for (size_t iv=0; iv < variables.size(); ++iv)
+            mol->set_geometry_variable(variables[iv].first, variables[iv].second);
+
+    } else {
+        std::vector <double> geom = molrec["geom"].cast<std::vector <double>>();
+        nat = geom.size() / 3;
+        unsettled = false;
+
+        for (size_t iat=0; iat<nat; ++iat) {
+            std::string symbol = elem[iat];
+            std::string label = elbl[iat];
+            std::transform(symbol.begin(), symbol.end(), symbol.begin(), ::toupper);
+            std::transform(label.begin(), label.end(), label.begin(), ::toupper);
+            mol->add_atom(elez[iat] * real[iat],
```

Comment:
Should we have more checks here to make sure say the lengths are the same? You can also use `eleze.at(iat)` for bounds check access.

Reply:
Thanks. `from_dict`s aren't revalidated (so hand-editing a molrec is not a safe operation), but it makes sense to add checks like this where easy.